### DISCUSSION TITLE
Unable to use GyroDriveBase

### DISCUSSION THREAD

**Discussion by mauroid** [pybricks/support#1068] [General]
Hi, this must be a novice question but..

This works fine:

from pybricks.robotics import DriveBase

But this produces a "from pybricks.robotics import GyroDriveBase" error:

from pybricks.robotics import GyroDriveBase

What am I doing wrong? I just copied the example from the doc, and changed Drivebase to GyroDriveBase, but can't compile. I am using v3.2.3 (Pybricks Code v2.1.1).

**Comment by laurensvalk** [comment]
Thanks for your message!

The gyro is a new feature that has only just been added to the beta version: https://beta.pybricks.com/

To try it, install the firmware from there, and write your code there, too.

Everything else should work the same.

We'd love to have your feedback on this feature. This will help us improve it before we include it in the official release too.

To see what is and isn't supported, just browse the documentation inside the app. This always matches the right version.

**Comment by ogzhnkse** [comment]
Hi, 
I have same error and I'm using beta version of Pybricks. Could you please help. 

Traceback (most recent call last):
  File "turkce.py", line 5, in <module>
ImportError: can't import name GyroDriveBase

My version:
('primehub', '3.6.0b2', 'ci-release-81-v3.6.0b2 on 2024-10-15')

**Reply by BertLindeman** [reply]
Hi,

IIRC the use of the gyro has now been integrated into the DriveBase.
import the DriveBase and tell it to use the gyro:

Example from the beta doc. The `use_gyro` command is commented out, change that and have fun:
```python
from pybricks.pupdevices import Motor
from pybricks.parameters import Port, Direction
from pybricks.robotics import DriveBase

# Initialize both motors. In this example, the motor on the
# left must turn counterclockwise to make the robot go forward.
left_motor = Motor(Port.A, Direction.COUNTERCLOCKWISE)
right_motor = Motor(Port.B)

# Initialize the drive base. In this example, the wheel diameter is 56mm.
# The distance between the two wheel-ground contact points is 112mm.
drive_base = DriveBase(left_motor, right_motor, wheel_diameter=56, axle_track=112)

# Optionally, uncomment the line below to use the gyro for improved accuracy.
# drive_base.use_gyro(True)

# Drive forward by 500mm (half a meter).
drive_base.straight(500)

# Turn around clockwise by 180 degrees.
drive_base.turn(180)

# Drive forward again to get back to the start.
drive_base.straight(500)

# Turn around counterclockwise.
drive_base.turn(-180)
``` 

Bert

**Reply by ogzhnkse** [reply]
Thanks Bert, that's actually how we were using it, but there is a slight deviation when going straight ahead after a turn, we're trying to figure out why.

**Reply by BertLindeman** [reply]
Might the new '3D calibration` in #1907 help you to drive better?

And / or do you have a small program to show what goes wrong?

Bert

**Comment by ogzhnkse** [comment]
As you can see in the video, after a 90-degree turn, there is a drift to the right. I’m also sending you the code. I couldn’t figure out what to write in the version field for the 3D calibration, so I couldn’t complete it. If you could help me with that, I would appreciate it, but this issue seems to be caused by something else. Thank you!
oguzhan

```python
from pybricks.hubs import PrimeHub
from pybricks.parameters import Direction, Port, Stop
from pybricks.pupdevices import Motor
from pybricks.robotics import DriveBase
from pybricks.tools import wait

prime_hub = PrimeHub()
right = Motor(Port.C, Direction.CLOCKWISE)
left = Motor(Port.B, Direction.COUNTERCLOCKWISE)
left_arm = Motor(Port.A, Direction.CLOCKWISE)
right_arm = Motor(Port.D, Direction.CLOCKWISE)
robot = DriveBase(left, right, 64.2, 117)


robot.settings(straight_acceleration=150)
robot.straight(300, Stop.COAST,True)
print(prime_hub.imu.heading())

robot.settings(turn_rate=1000)
robot.turn(95,Stop.COAST_SMART,True)
#robot.turn(93, Stop.BRAKE)
print(prime_hub.imu.heading())

robot.settings(straight_acceleration=150)
robot.straight(100, Stop.COAST,True)
print(prime_hub.imu.heading())
```

https://github.com/user-attachments/assets/ec4a20de-9cc8-492d-ad40-5faa1a4019ec

**Reply by BertLindeman** [reply]
Hi oguzhan,

The move seems only zero seconds long.
Could you retry to upload it?

Looking at the program now.

Bert

**Reply by dlech** [reply]
Firefox doesn't seem to like the video, but it works for me in other browsers and right-click "Save as..." also works to view it.

What firmware version are you using on the hub? Still the same as the earlier comment or one of the experimental versions?

**Reply by BertLindeman** [reply]
David, tried Edge, Firefox, Brave, Chrome to no avail. Maybe a problem with my windows. Oh, well..

**Comment by ogzhnkse** [comment]
It doesn’t look like that on my side. I’ve uploaded it again, and if it still shows as zero seconds, I’ll upload it to Drive.


https://github.com/user-attachments/assets/cd1736f9-352c-4f1d-bc7f-004f8d562f58

**Reply by BertLindeman** [reply]
Converted to mp4 after I downloaded from the link in the email.

https://github.com/user-attachments/assets/38795cef-eb98-412a-a88b-44b423fd71ae

Do  you mean the slight turn to the right directly after the turn?

I ran your test twice with the beta firmware on my primehub.
(The hub has not been calibrated.)

First test without use_gyro(True):
```
-0.190712
84.8444
87.05878
```
Then with  use_gyro(True):
```
-0.4267883
87.73534
94.50897
```

The heading of the second test is very near the 95 degrees of the turn.
I would advise using the beta firmware and  ```robot.use_gyro(True)```
To calibrate I would need to unbuild my model, as calibration is done with the hub without anything connected.
At the moment I did not have the time for that.

But calibration can improve precision.

Bert

**Reply by ogzhnkse** [reply]
Even trying this much means a lot to me, thank you so much. Here’s what I’ve tried so far:
	•	I attached Spike Prime large motors (the same issue persisted).
	•	I switched back to LEGO firmware and coded using word blocks. The deviation was very close, but not as much as with Pybricks.
	•	I tried again with Pybricks firmware and attached the Inventor medium motors (the same deviation issue occurred).
	•	I coded synchronized movement using motor blocks instead of the drive blocks (the issue still persisted).

It seems that if there’s a turn after the drive, there’s no issue with the drives, but if the turn comes before the drives, it appears to try to complete the missing degrees.

As you suggested, I’ll disassemble the hub and try calibration on a flat surface. I’ll report back with the results.

Oguzhan

**Reply by BertLindeman** [reply]
Hi Oguzhan,

If you are calibrating I would advice 3D calibration as Laurens explains in [the top post in issue 1907](https://github.com/pybricks/support/issues/1907)

Bert

**Reply by BertLindeman** [reply]
If you find it hard to choose which firmware to download:

The current PrimeHub nightly build firmware is [there](https://nightly.link/pybricks/pybricks-micropython/workflows/build/master/primehub-firmware-build-3649-git3b5326ab.zip)

**Reply by ogzhnkse** [reply]
Hello,
My wife woke up due to the beep sounds from the calibration and asked me to go to bed :)
I just completed the calibration, and the result is still the same; there’s a deviation after the turn. I’ll try it on a different robot today.
And yes I installed nightly version of firmware. 
Have great a day!
Oguzhan

**Reply by laurensvalk** [reply]
Thanks for the help everyone -- but I don't think we'll be needing any special firmwares for this :slightly_smiling_face: 

With `Stop.HOLD`, `Stop.BRAKE_SMART`, and  `Stop.COAST_SMART` (the one you are using), the robot will try to keep correcting errors across moves.

This helps you maintain targets across moves. So if your turn was a few degrees short due to friction for example, it will compensate as soon as it starts driving again. This is usually preferred. (If you don't want that though, you can use the normal `Stop.COAST`.)

Your design has quite a bit of weight away from the drive axle, so I'm not surprised it would miss a few degrees in turns. This is not necessarily an issue --- the compensation you are seeing is currently taking care of this.

Please feel free to open a new discussion for new questions. This seems unrelated to the outdated `GyroDriveBase` class. Thank you!
